File: apps/history-ingestor/history_ingestor/weeks.py
"""America/New_York calendar + ISO-week helpers.

The live SMA200W rule anchors the 199-week historical basis to the TRADING
WEEK of the latest quote, not to wall-clock "today" — so every timestamp must
be mapped to the NY week it belongs to. ISO weeks (Monday-based) are the
stable week identity: Alpha Vantage's weekly bucket dates are the week's
last trading day (a Thursday when Friday is a holiday, e.g. Good Friday
2025-04-17), and an ISO-week comparison keeps the anchor logic correct for
holiday weeks, early closes, weekends and year boundaries — no Friday-based
date arithmetic that would break on holiday weeks.
"""

from __future__ import annotations

import datetime as dt
from zoneinfo import ZoneInfo

NEW_YORK = ZoneInfo("America/New_York")

# Alpha Vantage WEEKLY bucket date keys are YYYY-MM-DD (week's last trading day).
DATE_KEY_RE = r"^\d{4}-\d{2}-\d{2}$"


def ny_date_of(instant: dt.datetime) -> dt.date:
    """NY calendar date for a UTC instant (DST-safe via zoneinfo)."""
    return instant.astimezone(NEW_YORK).date()


def iso_week_of(day: dt.date) -> tuple[int, int]:
    """(ISO year, ISO week) of a calendar date."""
    iso = day.isocalendar()
    return (iso[0], iso[1])


def iso_week_of_instant(instant: dt.datetime) -> tuple[int, int]:
    """(ISO year, ISO week) of the NY calendar date of a UTC instant."""
    return iso_week_of(ny_date_of(instant))


def week_label(day: dt.date) -> str:
    """Human/weekly-bucket identity, e.g. ``2026-W33``."""
    year, week = iso_week_of(day)
    return f"{year:04d}-W{week:02d}"


def week_label_of_date_key(date_key: str) -> str:
    """ISO week label of an Alpha Vantage date key (``YYYY-MM-DD``)."""
    return week_label(dt.date.fromisoformat(date_key))


def date_from_iso(date_key: str) -> dt.date:
    """Parse a ``YYYY-MM-DD`` key into a date (split/helper shared with the
    maintenance cycle)."""
    return dt.date.fromisoformat(date_key)


def is_in_progress_week(date_key: str, now: dt.datetime) -> bool:
    """Whether a weekly bucket belongs to the NY week that is still running
    at ``now`` (UTC). Buckets in the in-progress week are never stored as
    completed history — this keeps the historical basis deterministic no
    matter when bootstrap/maintenance runs, and guarantees the live formula
    never double-counts the quote's own week.
    """
    try:
        bucket_week = iso_week_of(dt.date.fromisoformat(date_key))
    except ValueError:
        return False  # malformed keys are rejected by the parser before this
    return bucket_week == iso_week_of_instant(now)


def completed_bars_filter(
    date_keys: list[str],
    now: dt.datetime,
) -> tuple[list[str], list[str]]:
    """Split bucket date keys into (completed, in_progress) by ``now``.

    Completed = strictly older NY week than the current one. The returned
    lists preserve input order.
    """
    completed: list[str] = []
    in_progress: list[str] = []
    for key in date_keys:
        if is_in_progress_week(key, now):
            in_progress.append(key)
        else:
            completed.append(key)
    return completed, in_progress


def target_completed_week(now: dt.datetime) -> str:
    """ISO week label of the last COMPLETED trading week as of ``now`` (UTC).

    The completed trading week is the one containing the most recent Friday
    (the market's weekly close) at or before today's NY date:
      Sunday  xxW34 (Fri xx-21)         -> "2026-W34"
      Monday  xxW35 (Fri xx-21 was W34) -> "2026-W34"
      next Sunday (Fri xx-28, W35)      -> "2026-W35"

    This is the maintenance cycle identity: Sunday STARTS a cycle for the week
    that closed on Friday (and reconciles its splits), and Monday's refresh
    stores exactly that same week's completed data.
    """
    ny = ny_date_of(now)
    days_since_friday = (ny.weekday() - 4) % 7  # Monday=0 .. Friday=4, Sat=5, Sun=6
    last_friday = ny - dt.timedelta(days=days_since_friday)
    return week_label(last_friday)


def is_monday_in_ny(now: dt.datetime) -> bool:
    """Whether ``now`` falls on a Monday in America/New_York.

    A weekly bucket only becomes storable (no longer the in-progress NY week)
    once Monday begins in NY, so the maintenance WEEKLY phase fetches only on
    NY Mondays. Sunday runs stay in the SPLITS phase.
    """
    return ny_date_of(now).weekday() == 0

File: apps/web/worker/sma/weeks.ts
/**
 * ISO-week helpers for the live SMA200W anchor logic.
 *
 * The live formula anchors the 199-week historical basis to the TRADING WEEK
 * of the latest quote (latest_quotes.provider_timestamp), never to wall-clock
 * "today". ISO weeks (Monday-based) are the stable week identity here:
 * Alpha Vantage weekly buckets are keyed by the week's last trading day (a
 * Thursday when Friday is a holiday, e.g. Good Friday 2025-04-17), and ISO
 * weeks keep the anchor comparison correct for holiday weeks, early closes,
 * weekends, Monday pre-open and year boundaries — no Friday-based date
 * arithmetic that would break on holiday weeks.
 *
 * Mirrors apps/history-ingestor/history_ingestor/weeks.py on purpose: the
 * Python side (maintenance) and this side (read path) must agree on what a
 * "week" is. Both compute the ISO week of an America/New_York calendar date.
 */

import { localNewYorkParts } from "../market-context";

export interface IsoWeek {
  year: number;
  week: number;
}

/**
 * ISO (year, week) of a calendar date (Gregorian, interpreted as a NY
 * calendar date). Thursday-based per ISO 8601.
 */
export function isoWeekOfDate(year: number, month: number, day: number): IsoWeek {
  const date = new Date(Date.UTC(year, month - 1, day));
  const weekday = date.getUTCDay(); // 0 = Sunday
  const isoWeekday = (weekday + 6) % 7; // 0 = Monday .. 6 = Sunday
  // Thursday of the same ISO week (may be in the previous calendar week for
  // Fri/Sat/Sun, hence a possibly-negative offset).
  const thursday = new Date(date);
  thursday.setUTCDate(date.getUTCDate() + (3 - isoWeekday));
  const thursdayYear = thursday.getUTCFullYear();
  // Monday of ISO week 1 of thursdayYear (Jan 4 is always in week 1).
  const jan4 = new Date(Date.UTC(thursdayYear, 0, 4));
  const week1Monday = new Date(jan4);
  week1Monday.setUTCDate(jan4.getUTCDate() - ((jan4.getUTCDay() + 6) % 7));
  const week = Math.floor((thursday.getTime() - week1Monday.getTime()) / (7 * 86_400_000)) + 1;
  return { year: thursdayYear, week };
}

/** ISO week of the NY calendar date of an instant (DST-safe). */
export function isoWeekOfNyInstant(instant: Date): IsoWeek | null {
  if (!Number.isFinite(instant.getTime())) return null;
  const parts = localNewYorkParts(instant);
  if (!parts) return null;
  return isoWeekOfDate(parts.year, parts.month, parts.day);
}

/** ISO week of an Alpha Vantage date key (YYYY-MM-DD). */
export function isoWeekOfDateKey(dateKey: string): IsoWeek | null {
  const match = /^(\d{4})-(\d{2})-(\d{2})$/.exec(dateKey);
  if (!match) return null;
  const year = Number(match[1]);
  const month = Number(match[2]);
  const day = Number(match[3]);
  const date = new Date(Date.UTC(year, month - 1, day));
  if (date.getUTCFullYear() !== year || date.getUTCMonth() !== month - 1 || date.getUTCDate() !== day) {
    return null; // impossible calendar date (e.g. 2026-02-30)
  }
  return isoWeekOfDate(year, month, day);
}

/** Monday (UTC midnight) of an ISO week — canonical week comparator. */
export function isoMonday(week: IsoWeek): Date {
  const jan4 = new Date(Date.UTC(week.year, 0, 4));
  const week1Monday = new Date(jan4);
  week1Monday.setUTCDate(jan4.getUTCDate() - ((jan4.getUTCDay() + 6) % 7));
  const monday = new Date(week1Monday);
  monday.setUTCDate(week1Monday.getUTCDate() + (week.week - 1) * 7);
  return monday;
}

/** Whole days between two ISO weeks: a - b (positive when a is later). */
export function weekDiffDays(a: IsoWeek, b: IsoWeek): number {
  return (isoMonday(a).getTime() - isoMonday(b).getTime()) / 86_400_000;
}
